### ISSUE TITLE
Dependency analysis partially broken

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#969]
Running the kernel extraction example (gocean/eg6) shows that loop boundaries are not detected as a read-dependency anymore, and therefore are missing from the written netcdf file. There seems to be two different(?) issues:
1. if a lower or upper bound is a derived type, this type is missing (do i=field%whole%xstart) Reason seems to be that the bound is a Literal, which does not provide variable access information. (I assume it should be a Reference???). Potential work-around/solution: in Literal.reference_accesses check if the string contains a '%'??? BAD hack, but it would make eg6 work again.
2. In similar cases I sometimes see undefined loop bounds as well. Not entirely sure when this happens exactly - I tried to fix 1) above, but now I get some variables correct, some as invalid:
```
      CALL extract_psy_data%PreStart("main", "init", 1, 6)
      CALL extract_psy_data%PreDeclareVariable("NOT_INITIALISED", NOT_INITIALISED)
...
      CALL extract_psy_data%PreDeclareVariable("b_fld%grid%dx", b_fld%grid%dx)

```
The latter is expected due to my bug fix for 1 ... but not sure about the first one

**Comment by hiker** [comment]
The other reason why I sometimes get undefined boundaries is actually caused by viewing a schedule: when viewing a schedule (before ``gen``), it will define the loop boundaries. So, when my script viewed the schedule, the boundaries where defined, if not, then the boundaries were undefined. So the real questions:

1. Can we set the boundaries earlier, i.e. when a loop object is created?
2. Otherwise I might need to explicitly set the loop boundaries when the ``reference_accesses`` is called. Though then the problem is that the functions that define the upper and lower loop boundaries are different in gocean and lfric.

**Comment by hiker** [comment]
While we have a testcase for loop boundaries handling, it is only with the NEMO API, which works (since it returns the upper loop boundary as a Reference, not as a Literal). I can also improve on the situation by creating a 'symbol' in _upper_limit, e.g.:
```
        stop_symbol = DataSymbol(stop, INTEGER_TYPE,
                                 interface=ArgumentInterface(
                                      ArgumentInterface.Access.READ))

        return Reference(stop_symbol)
        # instead of: return Literal(stop, INTEGER_TYPE, self)
```
This would then result in actually getting e.g. ``a_fld%whole%xstop`` as one variable with a read dependency, and while this is obviously not correct, it actually triggers the right behaviour: the name is then stored in the netcdf file, and properly handled by the driver:
```
      CALL extract_psy_data%ReadVariable("a_fld%whole%xstop", xstop_1)

```
Since we have several 'xstop' value, the driver does create unique scalars for each of them.

**Comment by hiker** [comment]
Re defining the loop limits earlier. I tried setting:
```
        self.start_expr = self._lower_bound()
        self.stop_expr = self._upper_bound()
```
at the end of the GOLoop constructor, but then (at least) ``self._const_loop_bounds`` is not defined, and later ``self._iteration_space`` is not defined either. Most of this can likely be all handled in ``GOKernCallFactory.create()`` (which will define iteration_space a bit later), The ``const_loop_bounds`` settings is adjusted later though (separate transform, so that would happen much later). I would suggest that upon setting ``const_loop_bounds``, we recompute the boundaries.

I wonder if something like ``const_loop_bounds` might be better of as a command line parameter instead of a transform (similar to distributed memory settings), this feels more intuitive to me (and would also solve that problem :) ).

**Comment by rupertford** [comment]
Would a variable in the gocean section of the config file be OK? If so, I would prefer that over the command line as it is not something that is applicable to to apis (not that dist mem is applicable to nemo :-)).

**Comment by hiker** [comment]
Good point - tbh, no urgency, and am happy to discuss this :) I just noticed that applying this as transformation (which actually only changes a flag) seems a bit unusual, compared with setting distributed memory ;) But no urgency for that.

**Comment by rupertford** [comment]
Yes, setting flags in transformations is not normal. We also do the same for inlining, but think the main difference there is that we might want to apply inlining selectively, whereas I think constant bounds is probably something that would be done for the whole code (although you never know with compilers!). @arporter, do you think we should keep the ability to selectively set the use of constant bounds? I can just imagine a case where a compiler works better for different options in the same code.

**Comment by arporter** [comment]
Yes, I can imagine that there may well be cases where we want to disable it for certain routines. Mind you, if we had it as a command-line parameter it could be enabled/disabled on a per-file basis which will probably be sufficient for the majority of cases. Unless it's going to cause real pain, I think I'd prefer to keep the current flexibility.

**Comment by sergisiso** [comment]
A few comments on this:

- Since yesterdary we support derived types, so now the Literls with '%' hack (e.g. "b_fld%grid%dx") can potentially be fixed to the appropriate reference. I can look into this. We also need derived types support in the back-end for this to work.

- I dissabled the constant loop bounds in gocean by default because I had some issues with it in NemoLite2D and it didn't seem to provide any performance advantage.

- I totally agree that the boundaries (and the whole PSyIR in fact) should be fully populated when the file is parsed (before calling the gen_code() routine). This is a requierement for using the PSyIR Back-ends in the PSy-layer and the back-ends should not contain any domain logic whatsoever. I will be doing this transition in gocean but this one may take longer to do.


Would tackling just the first bullet point for now be a solution?

**Comment by hiker** [comment]
I tried to find out details about this, but all I could see was a:
>     Some compilers are able to produce more efficient code if they are
>     provided with information on the relative trip-counts of the loops
>     within an Invoke. With constant loop bounds switched on, PSyclone

Does anyone remember details? I could check if there might be compiler options that can be added - am happy to study the assembly listing to find out what is wrong. IIrc, a Fortran do loop requires the relative trip-count to be known at the start of the loop (i.e. even if the upper boundary is a function call, it should only be called once at the start of the loop), so if using an upper bound from a struct makes a difference, that is somewhat odd. Then I would actually ask the other question: is there any disadvantage of creating scalars for the boundaries? 

> Yes, I can imagine that there may well be cases where we want to disable it for certain routines. Mind you, if we had it as a command-line parameter it could be enabled/disabled on a per-file basis which will probably be sufficient for the majority of cases. Unless it's going to cause real pain, I think I'd prefer to keep the current flexibility.

I find it hard to imagine that the same compiler creates better code for some loops with struct references, and for others with scalars only. I am happy with either solution that makes sure that the loop boundaries are defined when the loop is created. Implementing the const loop boundaries as transform means that the order of transformations can be important (well, of course in general the order of transformations is important, but in this case it seems odd). Just from my point of view - I would have thought that MPI should be a transform (since it modifies the tree), while this kind of change (which doesn't really change the tree, only the created code) should not be done as a transforms. Could this be perhaps a setting for the backend only (since Andy is concerned about generality ... I could imagine (maybe ;) ) that e.g. in the C version we might want to have const scalars as boundary, and in Fortran not. So, the loop should be specified with loop boundaries as the corresponding values from the field structures, and we could add this option to the backend maybe?? Though of course you can say that all the backend should do is convert the tree to the syntactically equivalent in the target language, and as such should not introduce new variables??

And while it is kind of stupid ... of course a setting in a config file can be used per-file ... just use different config files for different files. Yes, stupid idea, but considering that to me it seems extremely unlikely to ever need different behaviour for the same compiler, we could just go with the config file, and if this worst case really should happen, we have a workable solution.

I see a few solutions, and in decreasing preference:
1. Just always use const loop boundaries - and get rid of all the if statements and transforms.
2. Push it to the backend (though I am not certain about this).
3. Put it into the config file
4. Offer a command line option (this could obviously be combined with 3: get the default form config file, but command line option would overwrite this).

Notice ... using a transform to do this is imho not an option :)

**Comment by hiker** [comment]
> * Since yesterday we support derived types, so now the Literals with '%' hack (e.g. "b_fld%grid%dx") can potentially be fixed to the appropriate reference. I can look into this. We also need derived types support in the back-end for this to work.

That is indeed good news. 

> * I disabled the constant loop bounds in gocean by default because I had some issues with it in NemoLite2D and it didn't seem to provide any performance advantage.

OK, here I stand correct with my statement of "can there be any disadvantage of always using const loop boundaries" *sigh*

> * I totally agree that the boundaries (and the whole PSyIR in fact) should be fully populated when the file is parsed (before calling the gen_code() routine). This is a requirement for using the PSyIR Back-ends in the PSy-layer and the back-ends should not contain any domain logic whatsoever. I will be doing this transition in gocean but this one may take longer to do.

I give this thumbs up with all my hands and toes, and my family's hands and toes as well :)


> Would tackling just the first bullet point for now be a solution?

I think there is no immediate urgency to any of this (immediate as in: required for tutorial). Since I do psydata with LFRic, and we don't generate a driver yet, it is acceptable that something is missing (and I can have the faster participants try to find this bug :) ).

I actually need to take a step back to think about the whole concept, since in Iva's example I noticed for the first time that LFRic uses sometimes a function as upper limit 
```
      DO cell=1,perturbation_proxy%vspace%get_ncell()
        !
        CALL init_perturbation_code(nlayers, perturbation_proxy%data, chi_proxy(1)%data, chi_proxy(2)%data, chi_proxy(3)%data, &
&ndf_w3, undf_w3, map_w3(:,cell), ndf_wchi, undf_wchi, map_wchi(:,cell), basis_wchi_on_w3)
      END DO
````
Should we ask about performance issue caused by this :grin:? If accessing a struct element caused a slowdown, then a function call should be even worse.

But my issue with this is: even if the dependency analysis works perfectly with derived type and everything ... it would not return the information that a function call is being used (unless we say that the whole struct ``perturbation_proxy%vspace`` is read (which means we also need to have details about get_ncell to know that) ... and this then triggers the problem that we then should store the whole struct(?) ... it just gets pretty confusing, nothing I want to rush into providing a solution quickly tbh.

But in general, providing proper derived type info would be useful, and not affect me directly. We would nee to think what the dependency analysis should actually return. E.g.:
```
do i=1, f%upper_limit
```
Should ``reference_accesses()`` return that ``f%upper`` limit is read? Or should it return that ``f`` is (partially?) read? Or maybe even both? I am actually not sure what is best. 

Maybe we need to add a way to specify what it should do - e.g. the VariableAccesses instance that is passed in could have setting that specifies what exactly should be collected. So this could specify:
- to always return the lower and upper limit of loops, even if they are function calls
- what to do with derived type ... only include the master structure, only the actual variable in the structure? Both?

```
variables_info = VariablesAccessInfo(node_list, VariablesAccessInfo.COLLECT_LOOP_BOUNDS | ...)
```
Or:
```
variables_info = VariablesAccessInfo()
variables_info.set_collect_loop_bounds()
```
The actual syntax is flexible, but that sounds like a workable solution.

**Comment by arporter** [comment]
A couple of quick notes. It was with the Cray compiler that we saw a performance *advantage* with 'const loop bounds'. I don't think any other compiler (that we tried) managed to take advantage of it. It wasn't so much the removal of the reference to a component of a derived type; it was more that we could make it clear to the compiler how the limits on *all* of the loops in the invoke were related to each other (i.e. typically the same +/-1).

As to handling the dependence analysis for derived types, I think this is bound up in moving to using Symbols. The access will naturally be stored as a read access to a Symbol which then happens to be a ComponentSymbol etc. etc. Thanks to LFRic, I only recently learned that writing to an array pointed to from a component of derived type does *not* modify the derived type. Therefore, we'd need to know whether or not the ComponentSymbol was a pointer.
